ckan/tests/controllers/test_api.py

```diff
@@ -346,3 +350,32 @@ def test_i18n_only_known_locales_are_accepted(app):
     url = url_for("api.i18n_js_translations", ver=2, lang="unknown_lang")
     r = app.get(url, status=400)
     assert "Bad request - Unknown locale" in r.get_data(as_text=True)
+
+
+def yaml_serializer(data, headers, status):
+    return yaml.dump(data)
```

Comment:
this won't be able to handle objects with `for_json` methods so customizable serializers like this will "leak" into all our action calls by preventing their use.

Reply:
I think that it's the responsibility of the serializer's author to handle `for_json` usage.

CKAN core and any extension still can/should use `for_json` because it's a feature supported by the native serializer(not documented, but supported:)). Whenever an action returns a non-standard object(neither dictionary nor list, nor the other primitive), the native serializer "somehow" handles it. So whenever a developer registers a custom serializer, it's his responsibility to check the type of returned object and handle it properly.

As I understand, `for_json` is used because `json` library can handle only around 6 object types and everything else requires extra processing. If somebody uses `bson` or `toml` serializer, it gives him the support of additional types, such as DateTime, so the chances that `for_json` is needed are reduced. Some serializers, probably, allow you to use absolutely any type without preprocessing.

Other formats, like CSV, have almost no data types at all, so even `for_json` is not enough and the developer has to handle every returned value in a special way.

So I think that ability to set a custom serializer should not change anything in the existing code. It's just a possibility, with no guarantees.

BTW, it looks like `ckanapi` doesn't use `for_json` when serializing results during CLI invocation. Probably I should create an issue inside `ckanapi`, but I have no chance to verify whether this problem is real or I just missed something.

Reply:
The only user of `for_json` right now is the `datastore_search` api: https://github.com/ckan/ckan/blob/a18603ec182e279488c0e91b78cd957a3ed49d30/ckanext/datastore/backend/postgres.py#L1406

Using it improves performance when the final output of the call is going to be json. Code that treats this object like a list/dict should still work, the only problem I can see is with isinstance checks in some other serializer